cloud-sql-proxy/test_cloud_sql.py

```diff
@@ -0,0 +1,84 @@
+import unittest
+import datetime
+
+from parameterized import parameterized
+
+from integration_tests.dataproc_test_case import DataprocTestCase
+
+
+class CloudSqlProxyTestCase(DataprocTestCase):
+    COMPONENT = 'cloud-sql-proxy'
+    INIT_ACTION = 'gs://dataproc-initialization-actions/cloud-sql-proxy/cloud-sql-proxy.sh'
+    TEST_SCRIPT_LOCATION = 'dataproc-initialization-actions/cloud-sql-proxy'
+    TEST_SCRIPT_FILE_NAME = 'pyspark_metastore_test.py'
+    SCOPES = 'sql-admin'
+    DB_NAME = None
+
+    @classmethod
+    def setUpClass(cls):
+        super().setUpClass()
+        _, region, _ = cls.run_command("gcloud config get-value compute/region")
+        cls.REGION = region.strip() or "global"
+        _, project, _ = cls.run_command("gcloud config get-value project")
+        project = project.strip()
+        cls.PROJECT_METADATA = '{}:{}'.format(project, cls.REGION)
+
+    def setUp(self):
+        super().setUp()
+        self.DB_NAME = "test-{}-db".format(datetime.datetime.now().strftime("%Y%m%d%H%M%S"))
+        ret_code, stdout, stderr = self.run_command(
+            'gcloud sql instances create {} --region {}'.format(self.DB_NAME, self.REGION)
+        )
+        self.assertEqual(ret_code, 0, "Failed to create sql instance {}. Last error: {}".format(self.DB_NAME, stderr))
+
+    def tearDown(self):
+        super().tearDown()
+        ret_code, stdout, stderr = self.run_command(
+            'gcloud sql instances delete {}'.format(self.DB_NAME)
+        )
+        self.assertEqual(ret_code, 0, "Failed to delete sql instance {}. Last error: {}".format(self.DB_NAME, stderr))
+
+    def verify_instance(self, name):
+        self.__submit_pyspark_job(name)
+
+    def __submit_pyspark_job(self, name):
+        ret_code, stdout, stderr = self.run_command(
+            'gcloud dataproc jobs submit pyspark --cluster {} gs://{}/{}'.format(
+                name,
+                self.TEST_SCRIPT_LOCATION,
+                self.TEST_SCRIPT_FILE_NAME
+            )
+        )
+        self.assertEqual(ret_code, 0, "Failed to validate cluster. Last error: {}".format(stderr))
+
```

Comment:
We absolutely need read/write tests for Hive with an external table on GCS, since that's the main purpose of the init action:

```
gcloud dataproc clusters create <CLUSTER_NAME> \
--scopes sql-admin \
--initialization-actions gs://dataproc-initialization-actions/cloud-sql-proxy/cloud-sql-proxy.sh \
--properties hive:hive.metastore.warehouse.dir=gs://<HIVE_DATA_BUCKET>/hive-warehouse \
--metadata "hive-metastore-instance=<PROJECT_ID>:<REGION>:<INSTANCE_NAME>"
```

We should also have a test for KMS keys, e.g.:

```
gcloud dataproc clusters create <CLUSTER_NAME> \
--scopes cloud-platform \
--initialization-actions gs://<INIT_ACTION_BUCKET>/cloud-sql-proxy.sh \
--properties hive:hive.metastore.warehouse.dir=gs://<HIVE_DATA_BUCKET>/hive-warehouse \
--metadata "hive-metastore-instance=<PROJECT_ID>:<REGION>:<INSTANCE_NAME>" \
--metadata "kms-key-uri=projects/<PROJECT_ID>/locations/global/keyRings/my-key-ring/cryptoKeys/my-key" \
--metadata "db-admin-password-uri=gs://<SECRETS_BUCKET>/admin-password.encrypted" \
--metadata "db-hive-password-uri=gs://<SECRETS_BUCKET>/hive-password.encrypted"
```

In general, we should test at least all documented use cases. In an ideal world, we'd test all flags, even if they aren't documented.

Reply:
Thank you for the review, to be honest, I was just focused on verifying basic correctness of init cloud-sql-proxy, but I understood testing all cases makes a lot of sense. I will extend the testing script with missing scenarios and will let you know when ready.